Fatal Unhandled Exception when using Lambdas

This code works:
```
procedure foo() {
  var lam: [int]bool;
  lam := (lambda n: int :: n <= 0);
  assert lam[0]; 
} 
```
But this code:
```
procedure foo() {
  var lam: [int]bool where lam == (lambda n: int :: n <= 0);
  assert lam[0];
} 
```
crashes with
```
[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.Diagnostics.Contracts.ContractException: Assumption failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Diagnostics.Contracts.Contract.Assume (System.Boolean condition) [0x00003] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator+VariableMapping`1[VarKind].Lookup (VarKind boogieVar) [0x00009] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.LookupVariable (Microsoft.Boogie.Variable boogieVar) [0x00011] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitIdentifierExpr (Microsoft.Boogie.IdentifierExpr node) [0x00018] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.IdentifierExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitExpr (Microsoft.Boogie.Expr node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitBinderExpr (Microsoft.Boogie.BinderExpr node) [0x00008] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitLambdaExpr (Microsoft.Boogie.LambdaExpr node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.LambdaExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at VC.Wlp.Cmd (Microsoft.Boogie.Block b, Microsoft.Boogie.Cmd cmd, Microsoft.Boogie.VCExprAST.VCExpr N, VC.VCContext ctxt) [0x00477] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.Wlp.Block (Microsoft.Boogie.Block b, Microsoft.Boogie.VCExprAST.VCExpr N, VC.VCContext ctxt) [0x0003a] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.LetVCIterative (System.Collections.Generic.List`1[T] blocks, Microsoft.Boogie.VCExprAST.VCExpr controlFlowVariableExpr, System.Collections.Generic.Dictionary`2[TKey,TValue] label2absy, Microsoft.Boogie.ProverContext proverCtxt, System.Int32& assertionCount, System.Boolean isPositiveContext) [0x00270] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.GenerateVCAux (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VCExprAST.VCExpr controlFlowVariableExpr, System.Collections.Generic.Dictionary`2[TKey,TValue] label2absy, Microsoft.Boogie.ProverContext proverContext, System.Collections.Generic.IList`1[T] namedAssumeVars) [0x0012e] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen+Split.BeginCheck (Microsoft.Boogie.Checker checker, Microsoft.Boogie.VerifierCallback callback, VC.ModelViewInfo mvInfo, System.Int32 no, System.Int32 timeout) [0x00094] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00724] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
   --- End of inner exception stack trace ---
  at System.AggregateException.Handle (System.Func`2[T,TResult] predicate) [0x00064] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at Microsoft.Boogie.ExecutionEngine.InferAndVerify (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, System.String programId, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId) [0x004e0] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at Microsoft.Boogie.ExecutionEngine.ProcessFiles (System.Collections.Generic.List`1[T] fileNames, System.Boolean lookForSnapshots, System.String programId) [0x002bf] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at Microsoft.Boogie.OnlyBoogie.Main (System.String[] args) [0x0029d] in <a3e4aa56ba604f5fab835a20f4044358>:0 
---> (Inner Exception #0) System.Diagnostics.Contracts.ContractException: Assumption failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Diagnostics.Contracts.Contract.Assume (System.Boolean condition) [0x00003] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator+VariableMapping`1[VarKind].Lookup (VarKind boogieVar) [0x00009] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.LookupVariable (Microsoft.Boogie.Variable boogieVar) [0x00011] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitIdentifierExpr (Microsoft.Boogie.IdentifierExpr node) [0x00018] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.IdentifierExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitExpr (Microsoft.Boogie.Expr node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitBinderExpr (Microsoft.Boogie.BinderExpr node) [0x00008] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.VisitLambdaExpr (Microsoft.Boogie.LambdaExpr node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.LambdaExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <c38cfa7fb65449c4866e77cf76eb015c>:0 
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <40935da16c3347fdbd7383f3608d078e>:0 
  at VC.Wlp.Cmd (Microsoft.Boogie.Block b, Microsoft.Boogie.Cmd cmd, Microsoft.Boogie.VCExprAST.VCExpr N, VC.VCContext ctxt) [0x00477] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.Wlp.Block (Microsoft.Boogie.Block b, Microsoft.Boogie.VCExprAST.VCExpr N, VC.VCContext ctxt) [0x0003a] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.LetVCIterative (System.Collections.Generic.List`1[T] blocks, Microsoft.Boogie.VCExprAST.VCExpr controlFlowVariableExpr, System.Collections.Generic.Dictionary`2[TKey,TValue] label2absy, Microsoft.Boogie.ProverContext proverCtxt, System.Int32& assertionCount, System.Boolean isPositiveContext) [0x00270] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.GenerateVCAux (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VCExprAST.VCExpr controlFlowVariableExpr, System.Collections.Generic.Dictionary`2[TKey,TValue] label2absy, Microsoft.Boogie.ProverContext proverContext, System.Collections.Generic.IList`1[T] namedAssumeVars) [0x0012e] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen+Split.BeginCheck (Microsoft.Boogie.Checker checker, Microsoft.Boogie.VerifierCallback callback, VC.ModelViewInfo mvInfo, System.Int32 no, System.Int32 timeout) [0x00094] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00724] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <6966aa69d05e4fb3924ab9dc3762c064>:0 
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <cc441e1a3f6c4938becbc8567d722587>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 <---
```

--- a/Source/Core/LambdaHelper.cs
+++ b/Source/Core/LambdaHelper.cs
@@ -91,7 +91,7 @@ namespace Microsoft.Boogie
       }
     }
 
-    private class LambdaVisitor : StandardVisitor
+    private class LambdaVisitor : VarDeclOnceStandardVisitor
     {
       private readonly Dictionary<Expr, FunctionCall> liftedLambdas =
         new Dictionary<Expr, FunctionCall>(new AlphaEquality());

--- a/Source/Core/StandardVisitor.cs
+++ b/Source/Core/StandardVisitor.cs
@@ -818,6 +818,26 @@ namespace Microsoft.Boogie
   }
 
   /// <summary>
+  /// VarDeclOnceStandardVisitor is like StandardVisitor, except
+  ///   -- it does not visit a variable's declaration when visiting an `IdentifierExpr`, and
+  ///   -- it visits the `where` clause (if any) when visiting a `TypedIdent`.
+  /// </summary>
+  public abstract class VarDeclOnceStandardVisitor : StandardVisitor
+  {
+    public override Expr VisitIdentifierExpr(IdentifierExpr node) {
+      return node;
+    }
+
+    public override TypedIdent VisitTypedIdent(TypedIdent node) {
+      node = base.VisitTypedIdent(node);
+      if (node.WhereExpr != null) {
+        node.WhereExpr = (Expr) Visit(node.WhereExpr);
+      }
+      return node;
+    }
+  }
+
+  /// <summary>
   /// A ReadOnlyVisitor visits all the nodes of a given Absy.  The visitor may collect information from
   /// the nodes, may change fields contained in the data structure, but may not replace any nodes in the
   /// data structure.  To enforce this, all Visit...(node) methods have a postcondition that says that

--- a/Test/test2/LambdaWhere.bpl
+++ b/Test/test2/LambdaWhere.bpl
@@ -1,0 +1,28 @@
+// RUN: %boogie "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// These examples once crashed Boogie (see https://github.com/boogie-org/boogie/issues/111).
+
+procedure Local() {
+  var llam: [int]bool where llam == (lambda n: int :: n <= 0);
+  assert llam[0];
+}
+
+var glam: [int]bool where glam == (lambda n: int :: n <= 0);
+procedure Global() {
+  assert glam[0];
+}
+
+procedure InParameter(inlam: [int]bool where inlam == (lambda n: int :: n <= 0)) {
+  assert inlam[0];
+}
+
+procedure OutParameter() returns (outlam: [int]bool where outlam == (lambda n: int :: n <= 0)) {
+  assert outlam[0];
+}
+
+procedure ProcPlusImpl(inlam: [int]bool where inlam == (lambda n: int :: n <= 0))
+  returns (outlam: [int]bool where outlam == (lambda n: int :: n <= 0));
+implementation ProcPlusImpl(inlam: [int]bool) returns (outlam: [int]bool)
+{
+}

--- a/Test/test2/LambdaWhere.bpl.expect
+++ b/Test/test2/LambdaWhere.bpl.expect
@@ -1,0 +1,2 @@
+
+Boogie program verifier finished with 5 verified, 0 errors

